### ISSUE TITLE
define dockmaster.yml configuration file

### ISSUE THREAD

**Issue by briceburg** [dockerland/dockmaster#2]
Dockmaster requires a configuration file to interact with compostions. This file is the heart of dockmaster, and defaults to "dockmaster.yml" --  see #1 for how it is found. The configuration file provides orchestration instructions for a composition, such as; 
  * where it runs
  * the ability to render its docker-compose.yml file from templates
  * arbitrary execution of scripts through "hooks"
    *  such as assigning a network segment as part of a "pre" **up**  hook. 

#### where to run

Lets assign where a compostions runs through targets. Targets **must** specify a name and a type. Lets support the "docker-machine" type out-of-box with v1. 

```yml
---
targets:
  - name: node-a
    type: docker-machine

    # optional, defaults to name
    machine: node-a
```

> if multiple targets are defined, the compostion will be brought up/down from each. **TODO** allow flags to the dm command to specify targets by name

####  composition files

These directives should be _optional_ with dockmaster defaulting to using a docker-compose.yml file relative to the location of dockmaster.yml. Lets allow assigning an alternative location, and to support rendering of templates (using https://github.com/jasmith590/COP).
```yml
---
file:
  path: docker-compose.j2
  config:
    - /etc/dockmaster/defaults.yml
    - env.vars

#  ^^^ renders docker-compose.j2 using /etc/dockmaster/defaults.yml and ./env.vars
```

#### command hooks

Lets allow wrapping  _all_ commands with pre and post hooks. E.g. if `dm up -d` is executed, the command is `up`.   

Hooks should be thought of as "middleware", and allow manipulating of the arguments passed to the command (this may be super weird to do w/ shell... :cry: ), and halting execution on non-zero exit codes.  For v1 lets allow arbitrary v1 shell execution and script execution, For v2 lets explore built-in hooks via plugins (such as auto-detecting network segments for v2 compostions). 

```yml
----
commands:
  up:
    pre_hooks:
      - file: relative_path/to/hook.sh
      - file: /absolute/path/to/hook.sh
      - shell: rm -rf .working
```

> hooks are **always** executed relative to the configuration file 

#### arbitrary commands

Lets support non-native docker-compose commands, and use hooks for implementation.  E.g. `dm wiggle --jiggle trickle` (fires hooks for the `wiggle` command). To support this, we must not invoke docker-compose wiggle.

```yml
---
commands:
  wiggle:
    compose: false
    # ^^^ does not call docker-compose between pre- and post- hooks.
    pre_hooks:
      - file: wiggle.sh
```

### example
Putting this together, we get:

```yml
---
file:
  path: docker-compose.j2
  config:
    - /etc/dockmaster/defaults.yml
    - env.vars

targets:
  - name: node-a
    type: docker-machine

    # optional, defaults to name
    machine: node-a

commands:
  up:
    pre_hooks:
      - file: relative_path/to/hook.sh
      - file: /absolute/path/to/hook.sh
      - shell: rm -rf .working
  wiggle:
    compose: false
    # ^^^ does not call docker-compose between pre- and post- hooks.
    pre_hooks:
      - file: wiggle.sh
```

**Comment by samtay** [comment]
One small detail is that since we are allowing arbitrary commands, I think the commands should be an array instead of a map. Maybe thats pedantic, but I decided to at least try to parse this YML in haskell, and only did so after making [that change](https://github.com/SamTay/dockmaster/blob/39a86d6a21876f32dd7fa41a9ebda161c0d66606/test/fixtures/dockmaster.yml). (Of course we can marry these repos at some point if/when desired)

**Comment by briceburg** [comment]
@SamTay interesting. What is limiting it? This should be parsable. The nice thing about the map is you can't have a duplicate key (e.g. two definitions for up) Here's a few more ideas

``` yml
# your example, although has post_hooks
commands:
  - up:
      pre_hooks:
        - file: relative_path/to/hook.sh
        - file: /absolute/path/to/hook.sh
      post_hooks:
        - shell: rm -rf .working
```

``` yml
# adds 'name' key to the array of command maps
commands:
  - name: up
    pre:
      - file: pre-up.sh
    post:
      - shell: rm -rf .working
```

``` yml
# not sure how we can "intercept" compose on this one (e.g. to implement the wiggle command).
hooks:
  pre:
    - command: up
      file: pre-up.sh

    - command: up
      file: pre-up2.sh
```

We can merge in your dockmaster repo if you think it's ready. I was going to do a shell implementation first -- but if you want to go straight to haskell that's cool. Goal is to have a release _soon_ (just after Thanksgiving weekend?).

**Comment by samtay** [comment]
yeah you're right -- maybe I just got confused last night, sounds like map would be better to guarantee no dupes on commands. Nah I'd get the mvp in bash.. will probably end up grabbing snippets of it anyway

**Comment by chrispotter** [comment]
are these commands global to every time you run `dm` or `dmc` or are they going to be tethered to a specific composition? 
how will we specify the difference? 
pre-post hooks seems important to be tethered to the specific composition(at least the way i've been writing them thus far) they're instantiating (though there is an argument for general global commands like `dryrun` or `try` or something as well)

**Comment by briceburg** [comment]
@chrispotter 

> are these commands global to every time you run dm or dmc or are they going to be tethered to a specific composition? 

the dockmaster.yml describes a particular composition -- and my assumption has been that a dockmaster.yml is _required_ for `dm` to run.  The method of locating this file is defined in #1 

> how will we specify the difference? 

 I think we need to clear this up with verbiage -- as I'm assuming you're talking about the "compostion configuration file" (what this issue is all about),  and then the "global" or "user configuration file" we were thinking about defining the compostions paths in. 

For one, I'm thinking about using .ini for the "global|user configuration file", and yaml will be used for the "compostion configuration file". Here's the change to COP that allows .ini parsing; https://github.com/jasmith590/COP/commit/fd40391d2903b7fb9c1d1a646e172d5dfdf94411

Do you have a recommendation on what we should call these two files that will make immediate sense to everyone? 

> pre-post hooks seems important to be tethered to the specific composition

Yes they are -- as they're tied to the "compostion configuration file", aka dockmaster.yml as defined in this issue.

**Comment by briceburg** [comment]
OK -- @SamTay @chrispotter @Skrath @twslade @superterran @wrsenn are we in consensus on proposal 1 in first post?

**Comment by wrsenn** [comment]
👍 meets all my needs, at least.

**Comment by chrispotter** [comment]
@briceburg i think this clears up my concerns with 

> and my assumption has been that a dockmaster.yml is _required_ for dm to run

and I like the way this works out

**Comment by samtay** [comment]
Yup, think its the best & simplest one so far. Let's roll with it. I imagine we will tweak it a bit as we go.

**Comment by samtay** [comment]
@briceburg with multiple targets specified like below:
```yml
targets:
  - name: node-a
    type: docker-machine
  - name: node-b
    type: docker-machine

commands:
  - up:
      pre_hooks:
        - file: setup_machine.sh
        - file: only_run_once.sh
```
will we run hooks for each docker machine scope? That is, do we always do
```bash
(
  eval "$(docker-machine env node-a)"
  ./setup_machine.sh
  ./only_run_once.sh
  docker-compose up
)
(
  eval "$(docker-machine env node-b)"
  ./setup_machine.sh
  ./only_run_once.sh
  docker-compose up
)
```
and remove the possibility for a script that should only be run once, possibly outside of docker machine env subshell?

**Comment by briceburg** [comment]
@SamTay That is a great question, and I believe the proper way even though my gut feels like hooks fire once. 

Consider a pre-up hook that determines a network segment to assign to a composition -- it would need access to the machine to do this. Your scenario covers this, while this will not:

```sh
(
  ./setup_machine.sh
  ./only_run_once.sh

  for target in "${targets[@]"; do
    eval "$(docker-machine env $target)"
  done
)
```

We can perhaps come up with global single-fire hooks, or specify single-fire in the yaml (e.g. via a run_once boolean)

```yml
commands:
  - up:
      pre_hooks:
        - file: setup_machine.sh
        - file: only_run_once.sh
          run_once: True
```

although this may complicate things. For now lets implement firing of hooks per target.

**Comment by wrsenn** [comment]
Today's proposed changes:

```yml
---
compose:
  files:
    - path: docker-compose.yml
      type: file
    - path: docker-compose.j2
      type: template
      config:
        - fuck.vars
  flags:
    - "--project-name aaa"

env:
  files:
    - "somefile.env"
    - "somefile.json"
    - "somefile.xml"
    - "somefile.yml"
  vars:
    ROB_RULES: true

targets:
  - name: node-a
    type: docker-machine

    # optional, defaults to name
    machine: node-a

commands:
  up:
    pre_hooks:
      - file: relative_path/to/hook.sh
      - file: /absolute/path/to/hook.sh
      - shell: rm -rf .working
  wiggle:
    run_compose: false
    # ^^^ does not call docker-compose between pre- and post- hooks.
    pre_hooks:
      - file: wiggle.sh
  push:
    help: My help text
    run_compose: false
    pre_hooks:
      - shell: docker scp $1 $ROB_RULES nginx:/var/nginx/packages/
    post_hooks:
      - file: myuploadscript.sh
    

# dm push myfile.zip
```

**Comment by samtay** [comment]
Latest proposed dockmaster.yml has been implemented and merged (#7) to master.
